### PR TITLE
nxos_file_copy enhancement

### PR DESCRIPTION
**Pull request by saichint**
##### SUMMARY

Enhancement of nxos_file_copy module to include pulling a file from the device.

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

nxos_file_copy
##### ANSIBLE VERSION

```
ansible 2.7.0.dev0 (devel 313e7ba55e) last updated 2018/07/03 11:49:34 (GMT -400)
  config file = /root/agents-ci/ansible/test/integration/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /root/agents-ci/ansible/lib/ansible
  executable location = /root/agents-ci/ansible/bin/ansible
  python version = 2.7.6 (default, Oct 26 2016, 20:30:19) [GCC 4.8.4]

```

##### ADDITIONAL INFORMATION

* Currently nxos_file_copy module allows to copy a file from the ansible server to the device.
* For simultaneous transfers this may not be scalable, if there are too many devices and the file size is large (for ex. image bin file for the device)
* The proposed PR uses "copy scp" command to initiate file transfer from the device. 
* Since this needs passwords and can potentially go through multiple state machines, pexpect is used to traverse the states.
* Tested on all platforms and for the error conditions as well.

```

```

### CODE REVIEW

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (trishnaguha) [inline lib/ansible/modules/network/nxos/nxos_file_copy.py]: There is standard way to have these checks.
```
required_if = [("file_pull", True, ["remote_file", "remote_scp_server"]),
                      ("file_pull", False, ["local_file"])]

if file_pull: 
    check remote_file requirements
else:
    check_local_file requirements
```
Pass `required_if` to AnsibleModule(required_if=required_if) when creating the module object.

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (trishnaguha) [inline lib/ansible/modules/network/nxos/nxos_file_copy.py]: Can you please change the term `device` to `controller` here? Also in the doc.
Using device can be confusing here as it might mean network device.

**Review by trishnaguha — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (trishnaguha) [inline lib/ansible/modules/network/nxos/nxos_file_copy.py:373]: Why is `local_file` required here? Are `local_file` and `remote_file` not mutually exclusive?

File: test/integration/targets/nxos_file_copy/tests/cli/sanity.yaml
Comment (trishnaguha) [inline test/integration/targets/nxos_file_copy/tests/cli/sanity.yaml]: Is there any specific reason that the test is uncommented?

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (saichint) [reply lib/ansible/modules/network/nxos/nxos_file_copy.py:373]: No, remote_file and local_file are not mutually exclusive. Basically when we are copying files, if we want the destination file name to be same as source file name then we can provide just the remote_file. If we want the destination file name to be different, we can give local_file. So this is optional but not mutually exclusive.
The original code (for copying files using scp server on the switch) also has the same logic.

File: test/integration/targets/nxos_file_copy/tests/cli/sanity.yaml
Comment (saichint) [reply test/integration/targets/nxos_file_copy/tests/cli/sanity.yaml]: I believe you meant, why this test is ```commented``` out.
For file copy from the switch, we need to provide the remote_scp_server, remote_scp_username, remote_scp_password, remote_file_name etc. I tested this using my login credentials and the file_servers in our labs. We cannot put them in the test file as they are specific to our lab. So we cannot ```uncomment``` these tests until the real information of these parameters is given. I only put some example information there for now and commented it out.

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (mikewiebe) [inline lib/ansible/modules/network/nxos/nxos_file_copy.py:38]: I suggest modifying the `description:`  and `short_description` above this line to read as follows:
```
short_description: Copy a file to a NXOS device.
description:
  This module supports two different workflows for copying a file
  to flash (or bootflash) on NXOS devices.  Files can either be (1) pushed
  from the Ansible Server to the device or (2) pulled from a remote SCP
  file server to the device.  File copies are initiated from the NXOS
  device to the remote SCP server.  This module only supports the
  use of connection C(network_cli) or C(Cli) transport with connection C(local).

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (mikewiebe) [inline lib/ansible/modules/network/nxos/nxos_file_copy.py]: Please update the tested against version list to include what we validate in our matrix.

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (mikewiebe) [inline lib/ansible/modules/network/nxos/nxos_file_copy.py]: I would reword this section as follows:
```
- When pushing files (file_pull is False) to the NXOS device, feature scp-server must be enabled.
- When pulling files (file_pull is True) to the NXOS device, feature scp-server is not required.
- When pulling files (file_pull is True) to the NXOS device, no transfer will take place if the file is already present.

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (mikewiebe) [inline lib/ansible/modules/network/nxos/nxos_file_copy.py]: Does check mode work with both `file_pull: False` and `file_pull: True`?

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (mikewiebe) [inline lib/ansible/modules/network/nxos/nxos_file_copy.py]: IMO this might be more clear:
```
local_file:
  description:
    - When (file_pull is False) this is the path to the local file on the Ansible Server.
      The local directory must exist.
    - When (file_pull is True) this is the file name used rename the remote file when copied.
      This is optional when file_pull is True.
```

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (mikewiebe) [inline lib/ansible/modules/network/nxos/nxos_file_copy.py]: Likewise I would reword this to be:
```
remote_file
  description:
    - When (file_pull is False) this is remote file path on the NXOS device.
      If omitted, the name of the local file will be used.
      The remote directory must exist.
    - When (file_pull is True) this is the full path to the file on the remote SCP
      server to be copied to the NXOS device.

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (mikewiebe) [inline lib/ansible/modules/network/nxos/nxos_file_copy.py]: Please add `default: bootflash` for the `file_system:` parameter.

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (mikewiebe) [inline lib/ansible/modules/network/nxos/nxos_file_copy.py]: Suggest rewording as follows:
```
file_pull:
  description:
    - When (False) File is copied from the Ansible Server to the NXOS device.
    - When (True) File is copied from a remote SCP server to the NXOS device.
      In this mode, the file copy is initiated from the NXOS device.
    - If the file is already present on the device it will be overwritten and
      therefore the operation is NOT idempotent.

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (mikewiebe) [inline lib/ansible/modules/network/nxos/nxos_file_copy.py:111]: Please add specific examples for each of the following.

1. Transfer file from the Ansible server to the NXOS device.
2. Initiate file copy from the NXOS device to copy a file from the SCP server to the NXOS device.

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (mikewiebe) [inline lib/ansible/modules/network/nxos/nxos_file_copy.py]: What does `rt` stand for?  Might be good to have something a bit more descriptive.

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (mikewiebe) [inline lib/ansible/modules/network/nxos/nxos_file_copy.py]: Can u update the requirements to indicate which are needed for `file_pull: True` vs. `file_pull: False`?

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (mikewiebe) [reply lib/ansible/modules/network/nxos/nxos_file_copy.py:373]: Will this line throw an error when `file_pull` is `True` but `local_file` is not set in the playbook?

Comment (saichint) [reply lib/ansible/modules/network/nxos/nxos_file_copy.py:373]: No, because we are passing ```local``` which can be the name of ```local_file``` but if it is not in the playbook, it uses ```remote_file```

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (saichint) [reply lib/ansible/modules/network/nxos/nxos_file_copy.py]: yes

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (saichint) [reply lib/ansible/modules/network/nxos/nxos_file_copy.py:38]: done

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (saichint) [reply lib/ansible/modules/network/nxos/nxos_file_copy.py:111]: Done

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (saichint) [reply lib/ansible/modules/network/nxos/nxos_file_copy.py:38]: Done

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (saichint) [reply lib/ansible/modules/network/nxos/nxos_file_copy.py:111]: Done

**Comment by saichint** [comment]
@trishnaguha,
Can you please help me with the shippable errors for PR 
https://github.com/ansible/ansible/pull/42424

The main thing I noticed is that when we put
default: bootflash:
these problems are coming.
Basically Mike suggested to put the default value in the documentation but the default in the documentation is unable to take a colon at the end. If I remove it, it is not matching with the code and throws a different error. Do we know why the documentation does not accept a colon? If this cannot be resolved, please suggest any workaround or I can completely take it out and leave it as before.

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (mikewiebe) [inline lib/ansible/modules/network/nxos/nxos_file_copy.py]: Is the extra space here intentional?

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (mikewiebe) [inline lib/ansible/modules/network/nxos/nxos_file_copy.py]: Looks like I messed up the description for this... it should read:

`When (file_pull is True) this is the file name used when the remote file is copied from the SCP server to the NXOS device.`

Comment (mikewiebe) [reply lib/ansible/modules/network/nxos/nxos_file_copy.py]: What happens if (file_pull is True) and the `local_file` includes a directory path?

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (mikewiebe) [inline lib/ansible/modules/network/nxos/nxos_file_copy.py]: this is the

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (mikewiebe) [inline lib/ansible/modules/network/nxos/nxos_file_copy.py]: `# Initiate file copy from the nxos device to transfer file from an SCP server back to the nxos device.`

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (mikewiebe) [inline lib/ansible/modules/network/nxos/nxos_file_copy.py]: @saichint According to this 

https://stackoverflow.com/questions/8783705/how-could-should-i-state-colon-punctuation-in-a-yaml-file 

You should be able to use double quotes:

`default: "bootflash:"`

Comment (saichint) [reply lib/ansible/modules/network/nxos/nxos_file_copy.py]: I remember trying with quotes but did not work before. Let me check it again.

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (saichint) [reply lib/ansible/modules/network/nxos/nxos_file_copy.py]: Yes, this will have an issue if the local_file is anything other than a simple filename. Shall I add an error check if the local_file contains ```/``` ?

Comment (saichint) [reply lib/ansible/modules/network/nxos/nxos_file_copy.py]: I believe the same issue also happens in the original code if remote_file has directories. We can add error check there as well, let me know if checking for ```/``` is sufficient

**Comment by saichint** [comment]
@mikewiebe Thanks for your suggestion. The double quotes did the trick. I have finished work on your other comments except the local_file having directories. Once I get your reply, I will work on that

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (mikewiebe) [reply lib/ansible/modules/network/nxos/nxos_file_copy.py]: I would prefer the module try and create the full directory path if possible for the file transfer.

**Comment by mikewiebe** [comment]
👍

**Comment by saichint** [comment]
@trishnaguha After discussing with Mike, we decided to not include the full directory path for local file in this PR. I will work on that, once this gets merged because there is lot of design/code involved there. If you don't have any further comments, can you please merge it? Thanks

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (trishnaguha) [inline lib/ansible/modules/network/nxos/nxos_file_copy.py]: Please update this to "When (file_pull is True) this is the file name used on the NXOS device" .....

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (trishnaguha) [inline lib/ansible/modules/network/nxos/nxos_file_copy.py]: please add a local_file param example with file_pull being True

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (gundalow) [inline lib/ansible/modules/network/nxos/nxos_file_copy.py]: "Ansible Server" -> "Ansible controller"

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (gundalow) [inline lib/ansible/modules/network/nxos/nxos_file_copy.py]: suggestion: update message to include "when using file_pull"

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (gundalow) [inline lib/ansible/modules/network/nxos/nxos_file_copy.py]: suggestion: update message to include "when not using file_pull"

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (gundalow) [inline lib/ansible/modules/network/nxos/nxos_file_copy.py]: suggestion: update message to include "when not using file_pull"

File: test/integration/targets/nxos_file_copy/tests/cli/sanity.yaml
Comment (gundalow) [reply test/integration/targets/nxos_file_copy/tests/cli/sanity.yaml]: You could add `when: nxos_file_copy_server is defined`, so the tests are in, though only run in your local infra (and via host_var you can define these settings).

Other option would be to scp from localhost, or spin up a Python server on localhost (and a non-standard port, so it doesn't conflict with local sshd).

None of this needs to block this PR from progressing.

File: test/sanity/validate-modules/ignore.txt
Comment (gundalow) [inline test/sanity/validate-modules/ignore.txt:914]: Thanks :)

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (gundalow) [inline lib/ansible/modules/network/nxos/nxos_file_copy.py]: "Ansible Server" -> "Ansible controller"

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (gundalow) [inline lib/ansible/modules/network/nxos/nxos_file_copy.py]: "Ansible Server" -> "Ansible controller"

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (gundalow) [inline lib/ansible/modules/network/nxos/nxos_file_copy.py]: "Ansible Server" -> "Ansible controller"

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (gundalow) [reply lib/ansible/modules/network/nxos/nxos_file_copy.py]: "File" -> "file"

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (gundalow) [inline lib/ansible/modules/network/nxos/nxos_file_copy.py]: "File" -> "file"

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (gundalow) [inline lib/ansible/modules/network/nxos/nxos_file_copy.py:75]: Port of `I(remote_scp_server)` or ...?
Wonder if this should have an aliases `remote_scp_port`

**Review by gundalow — CHANGES_REQUESTED** [review]
When using a remote scp server will the module wait for the execution to be complete before returning, or is the transfer happening in the background?

File: test/integration/targets/nxos_file_copy/tests/cli/sanity.yaml
Comment (saichint) [reply test/integration/targets/nxos_file_copy/tests/cli/sanity.yaml]: I will take a look at this in the next iteration when I work on the local_file can include directory structure.

File: lib/ansible/modules/network/nxos/nxos_file_copy.py
Comment (saichint) [reply lib/ansible/modules/network/nxos/nxos_file_copy.py:75]: No, the connect_ssh_port is the ssh port of the network device. This is completely different from remote_scp information. The latter is a different machine which just runs the scp server. We do not need the scp port for that as of now because the ```copy scp ...``` command which is used on the device to pull the file from the remote scp server does not take any port information.

**Comment by saichint** [comment]
@trishnaguha @gundalow I fixed the the documentation according to your comments. The integration tests uncommenting part is something I will take a look in the next iteration as mentioned above. 
Thanks

UPDATE:
I was able to get the localhost to run as remote scp server and so I uncommented the tests. Now I am using the localhost as remote_scp_server and the ansible_ssh_user and ansible_ssh_pass as the remote_scp_server_user and remote_scp_server_password respectively. 
@trishnaguha Please try to run these tests in your lab and see if the tests run ok.

**Comment by saichint** [comment]
@gundalow Regarding your question about the background transfer, the current code waits for the execution to complete. This is why the timeout is a user parameter so that for large files or for low network speeds, this can be tuned properly.

**Review by trishnaguha — APPROVED** [review]

**Comment by saichint** [comment]
@trishnaguha could you merge this if you have no other comments? I am starting on local directory comment @mikewiebe asked me to, so it would be nice if this PR is merged so I can start clean on that,

**Pull request merged**
